torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -431,84 +431,189 @@ c10::optional<at::Tensor> ComputeConstantFolding(Node* n, int opset_version) {
+    if (shape[i] != 0) {
+      shape_ratio /= static_cast<uint64_t>(shape[i]);
+    } else {
+      shape_ratio /= static_cast<uint64_t>(input_shape[i]);
+    }
+  }
+
+  for (const int i : c10::irange(minus_one_pos)) {
+    int64_t cur_shape = shape[i] == 0 ? input_shape[i] : shape[i];
+    final_shape.push_back(cur_shape);
+  }
+  if (minus_one_pos != -1) {
+    final_shape.push_back(static_cast<int64_t>(std::round(shape_ratio)));
+  }
+  for (auto i = minus_one_pos + 1; i < shape_size; i++) {
+    int64_t cur_shape = shape[i] == 0 ? input_shape[i] : shape[i];
+    final_shape.push_back(cur_shape);
+  }
+  return final_shape;
+}
+
+// Similar to the function above, but for symbolic shapes.
+c10::optional<::c10::SymbolicShape> ComputeShapeFromReshape(
+    Node* n,
+    const c10::SymbolicShape& input_shape,
+    const c10::SymbolicShape& shape,
+    int opset_version) {
+  std::vector<c10::ShapeSymbol> input_shape_vector =
+      input_shape.sizes().value();
+  std::vector<c10::ShapeSymbol> shape_vector = shape.sizes().value();
+  TORCH_INTERNAL_ASSERT(
+      !input_shape_vector.empty() || !shape_vector.empty(),
+      "Reshape node should have at least one input size > 0 when constant folding.");
+  if (shape_vector.empty()) {
+    return input_shape;
+  }
+  if (input_shape_vector.empty()) {
+    return shape;
+  }
+
+  auto is_zero = [](c10::ShapeSymbol& ss) { return ss.value() == 0; };
+  auto it_0 = std::find_if(shape_vector.begin(), shape_vector.end(), is_zero);
+  bool shape_has_zero = it_0 != shape_vector.end();
+
+  if (opset_version >= 14 && n->hasAttributeS("allowzero")) {
+    int allowzero = n->i(attr::allowzero);
+    if (allowzero == 1 && shape_has_zero) {
+      // TODO: Shall we handle the case when shape has -1 here?
+      return shape;
+    }
+  }
+
+  int minus_one_pos = -1;
+  for (int i = 0; i < shape_vector.size(); ++i) {
+    if (shape_vector[i].value() == -1) {
+      minus_one_pos = i;
+      break;
+    }
+  }
+
+  if (!shape_has_zero && minus_one_pos == -1) {
+    return shape;
+  }
+  std::vector<c10::ShapeSymbol> final_shape;
+  uint64_t shape_ratio = 1.0;
+  std::unordered_map<int64_t, int64_t> sym_map;
+  for (const c10::ShapeSymbol& input_shape : input_shape_vector) {
+    if (input_shape.is_static()) {
+      shape_ratio *= static_cast<uint64_t>(input_shape.static_size());
+    } else {
+      auto value = input_shape.value();
+      sym_map.emplace(value, 0).first->second += 1;
+    }
+  }
+  int shape_size = static_cast<int>(shape_vector.size());
+  for (const int i : c10::irange(shape_size)) {
+    if (i == minus_one_pos) {
+      continue;
+    }
+    c10::ShapeSymbol& target_shape = shape_vector[i];
+    if (target_shape.value() == 0) {
+      target_shape = input_shape_vector[i];
+    }
+    if (target_shape.is_static()) {
+      shape_ratio /= static_cast<uint64_t>(target_shape.static_size());
+    } else {
+      auto value = target_shape.value();
+      if (sym_map.find(value) != sym_map.end()) {
```

Comment:
I think it'd be slightly more readable to handle `end()` case first, so we can remove a level of nesting.

```c++
if (sym_map.find(value) != sym_map.end()) {
  return c10::nullopt;
}
...
```